File: src/simulation/interactionmodel/MapModel.java
package simulation.interactionmodel;

import simulation.World;
import simulation.Event;
import simulation.culture.group.GroupConglomerate;

import java.util.*;

/**
 * Model with 2d map on which all interactions take place.
 */
public class MapModel implements InteractionModel {

    public List<Event> newEvents = new ArrayList<>();

    @Override
    public void turn(World world) {
        int size = world.groups.size();
        for (int j = 0; j < size; j++) {
            GroupConglomerate group = world.groups.get(j);
            int e = group.getEvents().size();
            group.update();
            for (int i = e; i < group.getEvents().size(); i++) {
                newEvents.add(group.getEvents().get(i));
            }
        }

        int a = world.events.size();
        world.map.update();

        for (GroupConglomerate group : world.groups) {
            group.finishUpdate();
        }

        world.map.finishUpdate();

        for (int i = a; i < world.events.size(); i++) {
            newEvents.add(world.events.get(i));
        }
    }

    @Override
    public void geologicTurn(World world) {
        int a = world.events.size();
        world.map.geologicUpdate();

        for (int i = a; i < world.events.size(); i++) {
            newEvents.add(world.events.get(i));
        }
    }

    @Override
    public Collection<Event> getEvents() {
        return newEvents;
    }

    @Override
    public void clearEvents() {
        newEvents = new ArrayList<>();
    }
}

File: src/simulation/interactionmodel/MapModel.kt
package simulation.interactionmodel

import simulation.Controller
import simulation.Event
import simulation.World

/**
 * Model with 2d map on which all interactions take place.
 */
class MapModel : InteractionModel {
    override var events: MutableList<Event> = mutableListOf()
    override val allEvents: MutableList<Event> = mutableListOf()

    override fun turn(world: World) {
        Controller.session.overallTime = System.nanoTime()
        Controller.session.groupTime = System.nanoTime()
        Controller.session.groupMainTime = 0
        Controller.session.groupOthersTime = 0
        Controller.session.groupInnerOtherTime = 0
        Controller.session.groupMigrationTime = 0

        val worldEventsNumber = world.events.size
        val groupEventSize = world.groups
                .map { it to it.events.size }
                .toMap()

        world.map.update()
        var groups = world.shuffledGroups
        for (group in groups) {
            group.update()
        }
        Controller.session.groupTime = System.nanoTime() - Controller.session.groupTime
        Controller.session.othersTime = System.nanoTime()

        groups = world.shuffledGroups
        for (group in groups)
            group.finishUpdate()
        world.map.finishUpdate()
        world.strayPlacesManager.update()

        events.addAll(world.events.drop(worldEventsNumber))
        allEvents.addAll(world.events.drop(worldEventsNumber))
        groups.forEach {
            events.addAll(it.events.drop(groupEventSize[it] ?: 0))
            allEvents.addAll(it.events.drop(groupEventSize[it] ?: 0))
        }

        Controller.session.othersTime = System.nanoTime() - Controller.session.othersTime
        Controller.session.overallTime = System.nanoTime() - Controller.session.overallTime

        val j = allEvents.groupBy { it }.filter { it.value.size > 1 }
        val i = events.groupBy { it }.filter { it.value.size > 1 }
        if (j.isNotEmpty()) {
            val k = 0
        }
        clearEvents()
    }

    override fun geologicTurn(world: World) {
        val worldEventsSize = world.events.size

        world.map.geologicUpdate()

        events.addAll(world.events.drop(worldEventsSize))
    }

    override fun clearEvents() = events.clear()
}
